### PR TITLE
3.next backports

### PR DESCRIPTION
**Pull request by dereuromark**
After diffing 4.next and 3.next, I found the following docblock and code changes that could be backported.

Some methods seem to be completely missing in API (validation), and here we would also need to backport tests then, of course.
https://github.com/cakephp/cakephp/pull/13538
=> I pulled those now into a separate PR https://github.com/cakephp/cakephp/pull/14682 for easier review.
Once that is done, I can rebase this one.

Note:
Validation has been fixed towards how 4.x behaves.
Now:

    $this->assertTrue(Validation::lengthBetween(1, 1, 3));

We should then also document this in the changelogs if we continue with it.

### CODE REVIEW

**Comment by ADmad** [comment]
This will again cause multiple conflicts when doing `3.next` to `master` merge :slightly_frowning_face:

**Comment by dereuromark** [comment]
I thought we didnt do that anymore? And thus so many changes here both ways anyway?

**Comment by ADmad** [comment]
Oh right, we only backport required changes. Sorry, I forgot.

**Comment by dereuromark** [comment]
I am still no git expert, but if the lines are exactly the same, on both sides, it should also go through without a conflict to resolve from my experience. But I could be wrong.

**Comment by markstory** [comment]
We haven't been doing merges from 3.next -> master in months. Doing backports by hand is a bit easier as we make fewer mistakes backporting incompatible typehints and don't have to deal with conflicts.

**Comment by othercorey** [comment]
The downside is 3.8 to 3.9 is going to be a huge diff with the formatting/doc changes we've made.

**Comment by dereuromark** [comment]
We can merge those first and i can reapply then these changes on top.

**Comment by dereuromark** [comment]
I rebased. btw, the newly added see links are not valid as they are ( https://github.com/cakephp/cakephp/pull/14679/commits/b018d3ab735e25a4b227d79d5aa9c4dc4c3baa09 ), missing the leading slash to work.

I am, however, a bit confused about the failing tests. I checked and cannot find the reason why any of the changes would affect the tests this way.

**Comment by dereuromark** [comment]
@othercorey Is master merged? And any idea what the failing test might be about?

**Comment by othercorey** [comment]
The failures are caused by the salt value. If you revert the change and return null when the Salt isn't set then it succeeds.

**Comment by dereuromark** [comment]
Do you think we should do that? Or would it be better to try to resolve the issue?
I am fine either way, but feel like that backporting 4.x could be the safer one here.

**Comment by dereuromark** [comment]
After your tip I found it looking into https://github.com/cakephp/cakephp/pull/12964
The tests were not fully correct without all changes, and I had to backport the whole PR here.

File: tests/TestCase/Auth/DigestAuthenticateTest.php [hunk 1]
```diff
@@ -61,12 +61,14 @@ public function setUp()
     {
         parent::setUp();
 
+        $salt = 'foo.bar';
+        Security::setSalt($salt);
```
Comment (othercorey) [inline tests/TestCase/Auth/DigestAuthenticateTest.php:65]: This would need to be reset although I'm not sure why it needs to be changed for this test.

Comment (dereuromark) [reply tests/TestCase/Auth/DigestAuthenticateTest.php:65]: That is a valid question. But we should check that on 4.x first, and if we find the reason to do so, I can update this PR.

Comment (dereuromark) [reply tests/TestCase/Auth/DigestAuthenticateTest.php]: You were right, in 4.next latest this is also not there anymore, so I cleaned this up now too.

**Pull request merged**